src/Microsoft.AspNetCore.Sockets/Internal/ConnectionState.cs

```diff
@@ -22,10 +24,25 @@ public ConnectionState(Connection connection, IChannelConnection<Message> applic
             LastSeenUtc = DateTime.UtcNow;
         }
 
-        public void Dispose()
+        public async Task DisposeAsync()
         {
+            // If the application task is faulted, propagate the error to the transport
+            if (ApplicationTask.IsFaulted)
+            {
+                Connection.Transport.Output.TryComplete(ApplicationTask.Exception.InnerException);
+            }
+
+            // If the transport task is faulted, propagate the error to the application
+            if (TransportTask.IsFaulted)
+            {
+                Application.Output.TryComplete(TransportTask.Exception.InnerException);
+            }
+
             Connection.Dispose();
             Application.Dispose();
+
+            // REVIEW: Add a timeout so we don't wait forever
+            await Task.WhenAll(ApplicationTask, TransportTask);
```

Comment:
I think this might be causing test hangs in my branch... if a channel is not drained the task will never complete. I am not sure if timeout is good enough.